### PR TITLE
make pool_size and overflow configurable for worker and portal

### LINKED ISSUE #192: QueuePool limit
I've recently upgraded from portal 2.x to 3.x using the docker deployment strategy.
After Uploading a couple of tracks and browsing the map the portal only returns errors. The reason is the set limit of the queue pool
```
Traceback (most recent call last):
  File "handle_request", line 62, in handle_request
    from sanic.models.futures import (
  File "_run_request_middleware", line 28, in _run_request_middleware
    Awaitable,
  File "/opt/obs/api/obs/api/app.py", line 182, in load_user
    await req.ctx.db.execute(select(User).where(User.id == user_id))
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/ext/asyncio/session.py", line 211, in execute
    return await greenlet_spawn(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 134, in greenlet_spawn
    result = context.throw(*sys.exc_info())
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 1691, in execute
    conn = self._connection_for_bind(bind)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 1532, in _connection_for_bind
    return self._transaction._connection_for_bind(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 747, in _connection_for_bind
    conn = bind.connect()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/future/engine.py", line 406, in connect
    return super(Engine, self).connect()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3204, in connect
    return self._connection_cls(self, close_with_result=close_with_result)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 96, in __init__
    else engine.raw_connection()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3283, in raw_connection
    return self._wrap_pool_connect(self.pool.connect, _connection)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3250, in _wrap_pool_connect
    return fn()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 310, in connect
    return _ConnectionFairy._checkout(self)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 868, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 476, in checkout
    rec = pool._do_get()
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/impl.py", line 134, in _do_get
    raise exc.TimeoutError(
sqlalchemy.exc.TimeoutError: QueuePool limit of size 5 overflow 10 reached, connection timed out, timeout 30.00 (Background on this error at: https://sqlalche.me/e/14/3o7r)
```

I guess the limit is too low. Can it be configured in the config.py or in the docker-compose.yml? It keeps crashing frequently in the current state. A restart of the portal cotainer only lasts for a couple of hours.

### PR DESCRIPTION
**Pull request by gluap**
Fixes #192

### CODE REVIEW

**Comment by opatut** [comment]
We should consider the underlying issue a bit more before blindly merging this. It might be a deeper problem. I think that the API should be able to handle a few parallel users with 10 connectinos and 30 seconds timeout, don't you? So maybe we're leaking connections and not returning them to the pool or something when an error occurs? Just raising the limit will not fix that.

We should also have a way for the API to recover from this problem. As far as I can tell, the error should cancel a single connection request, but if you wait for a bit for the pool to re-fill, the error should go away on its own. So this would be a temporary issue, one *not* requiring an API restart.

Now that I think about it, that *does* look like the connections are not returned properly. Please look into it a bit more. Set the limit very low, and check if you can recover from it after it has triggered.

**Comment by gluap** [comment]
Some of the experimenting VMs I use are ballooned out of memory at times and I wouldn't blink twice if I were to find that it can take them 30 seconds until a postgres has properly ballooned back and is happy to answer requests again. 

@SirTificate could you comment on the specs of the server where you encountered the issue? Is it a high performance system or do you feel that it is a bit on the edge at times?

@opatut I'd claim that having this configurable makes sense independently of whether the initial motivation to make it configurable is genuine need for more connections or need for connections caused by a bug - having it configurable does not have to keep us from looking for an underlying problem.

If the connections were not properly returned on a regular basis I'd assume that letting any instance run long enough would drive it into the issue. As far as I can tell the only longer-running instance currently operating is the one on portal.openbikesensor.org. If that one has not yet logged the issue, I'd assume that it can't be too easy to trigger.

**Comment by SirTificate** [comment]
The server I'm using is a Ubuntu 20.x VM with 2 Xeon Cores and 8GB of RAM. Hardware is enterprise grade, on fully redundant network storages. I can't see that the specs of the VM are a limiting factor. It actually runs pretty much idle.

However I think that @opatut has a valid point. If the Postgres DB itself is not limited by underperforming hardware the issue might lie deeper in the code. Unfortunately I'm not the right one to dig any deeper, as I don't have a full understanding of what's happening under the hood.
Raising the limit from 5 to 10 parallel connections however fixed the issue for our portal installation.

At the time I noticed the issues with the QueuePool I also had this issue https://github.com/openbikesensor/portal/issues/195, which happened to due a layer 8 error (me!). I have a gut feeling, that this might be related. Perhaps the missing road data caused an internal, but hidden, error which prevented the underlaying code from returning the database connection to the pool.

Whatever the reason might me, I still believe it's good to have the limit configurable. Especially since there are quite a lot of local OBS projects which are close to rolling out more hardware to more people. Sooner or later someone else will hit this limit again. It's just a matter of time!

**Comment by gluap** [comment]
> At the time I noticed the issues with the QueuePool I also had this issue #195, which happened to due a layer 8 error (me!). I have a gut feeling, that this might be related. Perhaps the missing road data caused an internal, but hidden, error which prevented the underlying code from returning the database connection to the pool.

@SirTificate that sounds plausible and may be a place to start the investigation.

In my esperience it is easy to miss exceptions in asyncio code and this may be a case where this happens.

**Comment by gluap** [comment]
I found a way to trigger the issue, very quickly zooming (as fast as possible) in and out using the scroll wheel does the trick. In case that matters: i also have a worker recomputing in the bg. Zooming on areas without map data seems to have no effect though.

Once even postgres crashef. 🤯

**Comment by opatut** [comment]
@gluap I sometimes restart(ed) portal.openbikesensor org to recover from this problem. It is indeed affected.

That shutdown of your postgres instance is concerning, though possibly unrelated. At least postgres has very good crash recovery ;)

**Comment by gluap** [comment]
I can now reproduce the original issue - setting overflow to 0 and pool size to 1 rather quickly triggers it.

Steps to reproduce:
- Overflow to 0, pool size to 1 (for this alone this PR is already useful ;)
- Zoom to map, zoom in and out a few times
- notice error message on console
- empty cache and hard reload on browser -> roads and measurements layer gone.
- Don't come back

Fixes I tried thus far:
- Guarding the database requests affected ``obs.api.routes.mapdetails``, ``obs.api.routes.tiles``  the error with try - except assuming that the exception might prevent returning of the connection to the pool - but this didn't help
- The issue seems related to the browser cancelling still-pending requests once the zoom level changes
- Through sanic and sqlalchemy this propagates towards postgres
- There was an [issue with sqlalchemy onche that roughly fit the description](https://github.com/sqlalchemy/sqlalchemy/issues/6652)  but it supposedly is fixed  in our version.

Currently looking into manually counting connections being checked out and returned to the pool in ``obs.api.db.make_session()´´

File: api/obs/api/db.py [hunk 1]
```diff
@@ -41,12 +41,12 @@
...
 
 engine = None
-sessionmaker = None
+sessionmaker: SessionMaker
 
 
 @asynccontextmanager
 async def make_session():
-    async with sessionmaker() as session:
+    async with sessionmaker(autoflush=True) as session:
```
Comment (gluap) [inline api/obs/api/db.py:49]: This allows the service to recover from the dead connections (or broken pool)

**Comment by gluap** [comment]
@opatut the latest commit (see comment) should at allow us to recover from the broken state - at least my formerly reproducing setup recovers from the issue with it.

Probably it only circumvents the underlying problem by renewing the broken connections from the pool. For fast scrolling I still see a few 503.

I'd suggest to leave the pool size and overflow customizable nonetheless.

**Comment by opatut** [comment]
I mean, it still hides the original problem, but if it works better in actual operation, let's use it for now. We probably won't fix it for real anytime soon then, but :shrug: 

Done is better than perfect :joy:

**Comment by opatut** [comment]
Thanks @gluap and @SirTificate for collaborating on this!

**Pull request merged**